core/src/main/java/com/linecorp/armeria/common/multipart/MultipartDecoder.java

```diff
@@ -0,0 +1,312 @@
+    @Nullable
+    private volatile Subscription upstream;
+    // Updated only via downstreamUpdater
+    @Nullable
+    private volatile Subscriber<? super BodyPart> downstream;
+
+    MultipartDecoder(String boundary) {
+        requireNonNull(boundary, "boundary");
+        parserEventProcessor = new ParserEventProcessor();
+        parser = new MimeParser(boundary, parserEventProcessor);
+        initFuture = new CompletableFuture<>();
+        bodyParts = new LinkedList<>();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super BodyPart> subscriber) {
+        requireNonNull(subscriber, "subscriber");
+        if (emitter != null || !downstreamUpdater.compareAndSet(this, null, subscriber)) {
+            subscriber.onSubscribe(SubscriptionHelper.CANCELLED);
+            subscriber.onError(new IllegalStateException("Only one Subscriber allowed"));
+            return;
+        }
+        deferredInit();
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (!upstreamUpdater.compareAndSet(this, null, subscription)) {
+            subscription.cancel();
+            throw new IllegalStateException("Subscription already set.");
+        }
+        deferredInit();
+    }
+
+    @Override
+    public void onNext(HttpData chunk) {
+        try {
+            final ByteBuf byteBuf;
+            if (chunk instanceof ByteBufHolder) {
+                byteBuf = ((ByteBufHolder) chunk).content();
+            } else {
+                byteBuf = Unpooled.wrappedBuffer(chunk.array());
+            }
+            parser.offer(byteBuf);
+            parser.parse();
+        } catch (MimeParsingException ex) {
+            emitter.fail(ex);
+            ReferenceCountUtil.release(chunk);
+        }
+
+        // submit parsed parts
+        while (!bodyParts.isEmpty()) {
+            if (emitter.isCancelled()) {
+                return;
+            }
+            emitter.emit(bodyParts.poll());
+        }
+
+        // complete the parts publisher
+        if (parserEventProcessor.isCompleted()) {
+            emitter.complete();
+            // parts are delivered sequentially
+            // we potentially drop the last part if not requested
+            emitter.clearBuffer(MultipartDecoder::drainPart);
+        }
+
+        // request more data to detect the next part
+        // if not in the middle of a part content
+        // or if the part content subscriber needs more
+        if (upstream != SubscriptionHelper.CANCELLED &&
+            emitter.hasRequests() &&
+            parserEventProcessor.isDataRequired() &&
+            (!parserEventProcessor.isContentDataRequired() || bodyPartPublisher.hasRequests())) {
+            upstream.request(1);
+        }
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        requireNonNull(throwable, "throwable");
+        initFuture.whenComplete((e, t) -> e.fail(throwable));
```

Comment:
`initFuture` does not complete with an exception. How about using `thenApply`?

Reply:
Let me use `thenAccept` :-)